packages/react-tooltip/Spec.md

```diff
@@ -1,63 +1,474 @@
+
+```typescript
+/**
+ * Create a ref that, when attached to an element, shows the tooltip on hover or focus.
+ *
+ * @param tooltip - The tooltip to display on hover or focus. Can be a string, JSX element tree, or TooltipProps.
+ */
+export function useTooltipRef(tooltip: ShorthandProps<TooltipProps>);
+```
+
+## @fluentui/react-tooltip
+
+The `react-tooltip` package contains the bulk of the implementation of tooltips, including rendering, styling, positioning, lifetime management, etc. This will be a larger package, but can be lazy-loaded and doesn't need to be included in an app's bundle directly.
+
+### Tooltip
+
+`Tooltip` renders the tooltip content itself, and the arrow that points to the target element.
+
+The `TooltipProps` interface is defined in the `react-tooltip-provider` package.
+
+The `TooltipState` type has a slot for the arrow. This arrow slot can't be set by a prop, but is used internally for rendering and positioning.
+
+```typescript
+import { TooltipProps } from '@fluentui/react-tooltip-provider';
+
+export { TooltipProps };
+
+export const tooltipShorthandProps = [] as const;
+export const tooltipSlotProps = [...tooltipShorthandProps, 'arrow'] as const;
+
+export type TooltipState = ComponentState<
+  TooltipProps & {
+    /**
+     * The arrow that points to the target element.
+     */
+    arrow?: ObjectShorthandProps<React.HTMLAttributes<HTMLElement> & React.RefAttributes<HTMLElement>>;
+  },
+  /* ShorthandProps: */ typeof tooltipShorthandProps[number],
+  /* DefaultedProps: */ 'arrow' | 'placement'
+>;
+```
+
+### TooltipManager
+
+`TooltipManager` handles showing and hiding tooltips, including positioning and tracking the visible tooltip. There is only one instance at the root of the app.
+
+Note that the `TooltipManager` has a slot for the visible tooltip on its State, but it's not a public prop. Instead, the tooltip is shown/hidden using the `TooltipManagerApi` imperative interface, which is defined in the `react-tooltip-provider` package below.
+
+```typescript
+import { TooltipManagerApi } from '@fluentui/react-tooltip-provider';
+
+export interface TooltipManagerProps extends ComponentProps, React.HTMLAttributes<HTMLElement> {
+  /**
+   * Ref to the imperative interface to show and hide tooltips.
+   */
+  componentRef?: React.Ref<TooltipManagerApi>;
+}
+
+export const tooltipManagerShorthandProps = ['tooltip'] as const;
+
+export type TooltipManagerState = ComponentState<
+  TooltipManagerProps & {
+    /**
+     * The Tooltip being rendered, if any.
+     */
+    tooltip?: ShorthandProps<TooltipProps>;
+  }
+>;
+```
+
+# Structure
+
+## Public
+
+```tsx
+<TooltipProvider>
+  <Button tooltip="Hello world" /> {/* for example... */}
+</TooltipProvider>
+```
+
+## Internal
+
+`TooltipProvider`:
+
+```tsx
+<internal__TooltipProviderContext.Provider value={{ current: undefined }}>
+  <slots.root {...slotProps.root} />
+  <React.Suspense fallback={null}>
```

Comment:
I have heard different things about supporting server side rendering so I do not know for sure if it is a requirement. But I think `Suspense` is not available for SSR: https://reactjs.org/docs/code-splitting.html#reactlazy

Reply:
Yeah. The lazy-loading aspect is not necessarily a hard requirement of this design.

The bigger reason for splitting out the react-tooltip-provider and react-tooltip packages is so that e.g. `Button` can have a `tooltip` prop, but not have a direct dependency on react-tooltip. That way, if someone wants Button but doesn't need tooltips, they don't pay the cost of the react-tooltip package.